pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,446 @@
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+// recvPing receives an ICMP message from the target IP address.
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool) {
+	// We only expect small packets, if we receive a larger packet, we will drop the extra data.
+	readBuffer := make([]byte, 128)
+	for {
+		// If the socket is closed, it will return
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			klog.ErrorS(err, "Failed to read ICMP message")
+			return
+		}
+
+		destIP := peer.String()
+
+		// Parse the ICMP message
+		var msg *icmp.Message
+		if isIPv4 {
+			msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv4.ICMPTypeEchoReply {
+				klog.V(4).InfoS("Failed to match ICMPTypeEchoReply")
+				continue
+			}
+		} else {
+			msg, err = icmp.ParseMessage(ProtocolICMPv6, readBuffer)
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv6.ICMPTypeEchoReply {
+				klog.V(4).InfoS("Failed to match ICMPTypeEchoReply")
+				continue
+			}
+		}
+
+		echo, ok := msg.Body.(*icmp.Echo)
+		if !ok {
+			klog.ErrorS(nil, "Failed to assert type as *icmp.Echo")
+			continue
+		}
+
+		klog.V(4).InfoS("Recv ICMP message", "IP", destIP, "echo", echo)
+
+		// Parse the time from the ICMP data
+		sentTime, err := time.Parse(time.RFC3339Nano, string(echo.Data))
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse time from ICMP data")
+			continue
+		}
+
+		// Calculate the round-trip time
+		end := time.Now()
+		rtt := end.Sub(sentTime)
+
+		// Update the latency store
+		mutator := func(entry *NodeIPLatencyEntry) {
+			entry.LastSendTime = sentTime
+			entry.LastRecvTime = end
+			entry.LastMeasuredRTT = rtt
+		}
+		m.latencyStore.SetNodeIPLatencyEntry(destIP, mutator)
+	}
+}
+
+// pingAll sends ICMP messages to all the Nodes.
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	for _, toIP := range nodeIPs {
+		if toIP.To4() != nil && ipv4Socket != nil {
+			if err := m.sendPing(ipv4Socket, toIP); err != nil {
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Failed to send ICMP message to Node")
+		}
+	}
+	klog.V(4).InfoS("Ping all Nodes done")
```

Comment:
```suggestion
	klog.V(4).InfoS("Done pinging all Nodes")
```

What would be nice is to have a corresponding log message at the beginning of the function: `klog.V(4).InfoS("Pinging all Nodes")`

Reply:
That make sense, I will change it.